### PR TITLE
[Inductor] CPP Backend Tiling Abstraction

### PR DESCRIPTION
**Pull request by jgong5**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96407
* __->__ #95585

This PR aims to provide an extensible infrastructure for CPP codegen supporting loop tiling and optimal tile-based ops leveraging language/library intrinsics. The generated code matches those generated by the current master except for the one generated by CppTile2DTailKernel (explained later in details).

### Summary of changes
* A `CppTileKernel` class accompanied by `CppTileOverrides` is added to operate on n-D tiles with fixed sizes by tiling “n” kernel loop levels with these sizes. It is a generalization of existing `CppVecKernel` (1D tiles with tile size 8 or 16) and `CppTile2DKernel` (2D tiles with tile sizes 8x8 or 16x16). It deprecates `CppVecKernel`, `CppTile2DKernel` and `CppTile2DTailKernel`.
* n-D tile variable is modeled with `CppTileCSEVariable` (a subclass of `CSEVariable`) with `TileMeta` as attribute to describe tile meta info: dtype, sizes, tile axes, data layout and the inner-loop scope the tile is executed.
* Ops dispatching (`CppTileKernel.dispatch_ops` and `CppTileKernel.tile_ops_tbl`) based on tile ranks with a fallback chain is implemented to lower the ops to leverage language/library intrinsics as much as possible (`CppTile2DOverrides` for 2D transpose, `CppTile1DOverrides` for 1D vec ops, `CppTile0DOverrides`), or if intrinsics not available, fall back to the loop-based implementation around the “rank minus 1” handler (`CppTileFallbackOverrides` and `CppTileFallbackWrapper`).
* `CppTileCodeGenBuffer` overloading `IndentedBuffer` to record code lines with inner loop scope info, helps inner loop fusion, tile buffer management and final code cleanup before dumping the final kernel code.

Arch diagram is shown below:
image

### Some notes on design details
1. Tile buffers: ops from `CppTileKernel` accept and return the full-rank tiles while the actually computation might happen on a slice of the tile from the inner loop scope, for which case, a tile buffer will be defined to make sure the tile variable is still live to be used by later ops outside the loop scope. A `tile_slice_store` will be executed after each definition of those tile slices. The uses of tile buffers are tracked by `TileBufferUseTracker`. If the tile buffer is not used by later ops, it will be removed by the final codegen.
2. Inner loop fusion: the final codegen inside `CppTileKernel.__exit__` scans the `CppTileCodeOrLine` from `CppTileCodeGenBuffer` and tries to fuse those code lines into the same inner loop level if they belong to the same loop scope (i.e., with same `compute_at` value).

### Generated code example
Generated code from 1D vectorized code on inner-most loop is kept the same. Below is the code generated by tile 2d with tile2d transposition, generated from [`test_transpose_add`](https://github.com/pytorch/pytorch/blob/7a192cc51c7172860efd35413acf9a8b9aafd2c9/test/inductor/test_torchinductor.py#L2990). There are three kernel instances generated for this case. The first two are generated by `CppTileKernel` (originally they are generated by `CppTile2DKernel` and `CppTile2DTailKernel` respectively) and the last one is generated by `CppKernel` as before. The first kernel has 2d tiling 16x16 on two loop levels. The code pattern is almost the same as those from github master except that tile variables take `std::array` as the data type. The second kernel has 1d tiling on the outer loop level. Different from master which generates pure scalar code, we are now generating vectorized code as much as possible and fall back to loop-based scalar code if necessary (here, we fall back to the scalar load since it is non-contiguous).

Before
```c++
#include "/tmp/torchinductor_root/dl/cdljpywww2h2ag4o35mwbvm45hhasxnxkhqgbupxnk3y7olula65.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       const float* __restrict__ in_ptr1,
                       float* __restrict__ out_ptr0)
{
    #pragma omp parallel num_threads(64)
    {
        {
            #pragma omp for 
            for(long i0=0; i0<2; i0+=1)
            {
                #pragma GCC ivdep
                for(long i1=0; i1<1; i1+=1)
                {
                    float tmp0[16*16] __attribute__ ((aligned (16)));
                    at::vec::transpose_mxn<float,16,16>(in_ptr0 + (16*i0) + (512*i1), 32, tmp0, 16);
                    float tmp4[16*16] __attribute__ ((aligned (16)));
                    for (long i0_inner = 0; i0_inner < 16; i0_inner++)
                    {
                        auto tmp1 = at::vec::Vectorized<float>::loadu(tmp0 + 16*i0_inner);
                        auto tmp2 = at::vec::Vectorized<float>::loadu(in_ptr1 + (16*i0_inner) + (16*i1) + (256*i0));
                        auto tmp3 = tmp1 + tmp2;
                        tmp3.store(tmp4 + 16*i0_inner);
                    }
                    at::vec::transpose_mxn<float,16,16>(tmp4, 16, out_ptr0 + (16*i0) + (512*i1), 32);
                }
                #pragma GCC ivdep
                for(long i1=16; i1<16; i1+=1)
                {
                    for (long i0_inner = 0; i0_inner < 16; i0_inner++)
                    {
                        auto tmp0 = in_ptr0[i0_inner + (16*i0) + (32*i1)];
                        auto tmp1 = in_ptr1[i1 + (16*i0_inner) + (256*i0)];
                        auto tmp2 = tmp0 + tmp1;
                        out_ptr0[i0_inner + (16*i0) + (32*i1)] = tmp2;
                    }
                }
            }
            #pragma omp for 
            for(long i0=32; i0<32; i0+=1)
            {
                #pragma GCC ivdep
                for(long i1=0; i1<16; i1+=1)
                {
                    auto tmp0 = in_ptr0[i0 + (32*i1)];
                    auto tmp1 = in_ptr1[i1 + (16*i0)];
                    auto tmp2 = tmp0 + tmp1;
                    out_ptr0[i0 + (32*i1)] = tmp2;
                }
            }
        }
    }
}
```

After
```c++
#include "/tmp/torchinductor_root/dl/cdljpywww2h2ag4o35mwbvm45hhasxnxkhqgbupxnk3y7olula65.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       const float* __restrict__ in_ptr1,
                       float* __restrict__ out_ptr0)
{
    #pragma omp parallel num_threads(64)
    {
        {
            #pragma omp for 
            for(long i0=0; i0<2; i0+=1)
            {
                #pragma GCC ivdep
                for(long i1=0; i1<1; i1+=1)
                {
                    __at_align__ std::array<float,256> tmp5;
                    auto tmp0 = ([&]() { __at_align__ std::array<float,256> tmp; at::vec::transpose_mxn<float,16,16>(in_ptr0 + (16*i0) + (512*i1), 32, &tmp[0], 16); return tmp; })();
                    for (long i0_inner = 0; i0_inner < 16; i0_inner++) {
                        auto tmp1 = at::vec::Vectorized<float>::loadu(&in_ptr1[(16*i0_inner) + (16*i1) + (256*i0)]);
                        auto tmp3 = at::vec::Vectorized<float>::loadu(&tmp0[16*i0_inner]) /* slice load */;
                        auto tmp4 = tmp3 + tmp1;
                        tmp4.store(&tmp5[16*i0_inner]) /* slice store */;
                    }
                    at::vec::transpose_mxn<float,16,16>(&tmp5[0], 16, out_ptr0 + (16*i0) + (512*i1), 32);
                }
                #pragma GCC ivdep
                for(long i1=16; i1<16; i1+=1)
                {
                    __at_align__ std::array<float,16> tmp2;
                    auto tmp0 = at::vec::Vectorized<float>::loadu(&in_ptr0[(16*i0) + (32*i1)]);
                    for (long i0_inner = 0; i0_inner < 16; i0_inner++) {
                        auto tmp1 = in_ptr1[i1 + (16*i0_inner) + (256*i0)];
                        tmp2[i0_inner] = tmp1 /* slice store */;
                    }
                    auto tmp3 = at::vec::Vectorized<float>::loadu(&tmp2[0]) /* register load */;
                    auto tmp4 = tmp0 + tmp3;
                    tmp4.store(out_ptr0 + (16*i0) + (32*i1));
                }
            }
            #pragma omp for 
            for(long i0=32; i0<32; i0+=1)
            {
                #pragma GCC ivdep
                for(long i1=0; i1<16; i1+=1)
                {
                    auto tmp0 = in_ptr0[i0 + (32*i1)];
                    auto tmp1 = in_ptr1[i1 + (16*i0)];
                    auto tmp2 = tmp0 + tmp1;
                    out_ptr0[i0 + (32*i1)] = tmp2;
                }
            }
        }
    }
}
```
Below is the generated code if we don't apply the tile slice load cache. By default, we always pass the tile buffers at the outer-most scope between tile ops. Every time a new tile variable is defined from an inner loop scope, a tile slice store is executed to the corresponding tile buffer at the outer-most scope to make sure the tile buffer is up-to-date. Then, the tile buffer is sliced per the inner loop scope before being passed to other users. With tile slice load cache, the extra slice loads are saved (e.g., `tmp4`) and then, the tile buffers (e.g., `tmp2`) could be deleted (by the final codegen in `CppTileKernel.__exit__`) if they are not used in other places at all.
```c++
#include "/tmp/torchinductor_root/dl/cdljpywww2h2ag4o35mwbvm45hhasxnxkhqgbupxnk3y7olula65.h"
extern "C" void kernel(const float* __restrict__ in_ptr0,
                       const float* __restrict__ in_ptr1,
                       float* __restrict__ out_ptr0)
{
    #pragma omp parallel num_threads(64)
    {
        {
            #pragma omp for 
            for(long i0=0; i0<2; i0+=1)
            {
                #pragma GCC ivdep
                for(long i1=0; i1<1; i1+=1)
                {
                    __at_align__ std::array<float,256> tmp2;
                    __at_align__ std::array<float,256> tmp6;
                    auto tmp0 = ([&]() { __at_align__ std::array<float,256> tmp; at::vec::transpose_mxn<float,16,16>(in_ptr0 + (16*i0) + (512*i1), 32, &tmp[0], 16); return tmp; })();
                    for (long i0_inner = 0; i0_inner < 16; i0_inner++) {
                        auto tmp1 = at::vec::Vectorized<float>::loadu(&in_ptr1[(16*i0_inner) + (16*i1) + (256*i0)]);
                        tmp1.store(&tmp2[16*i0_inner]) /* slice store */;
                        auto tmp3 = at::vec::Vectorized<float>::loadu(&tmp0[16*i0_inner]) /* slice load */;
                        auto tmp4 = at::vec::Vectorized<float>::loadu(&tmp2[16*i0_inner]) /* slice load */;
                        auto tmp5 = tmp3 + tmp4;
                        tmp5.store(&tmp6[16*i0_inner]) /* slice store */;
                    }
                    at::vec::transpose_mxn<float,16,16>(&tmp6[0], 16, out_ptr0 + (16*i0) + (512*i1), 32);
                }
                #pragma GCC ivdep
                for(long i1=16; i1<16; i1+=1)
                {
                    __at_align__ std::array<float,16> tmp2;
                    auto tmp0 = at::vec::Vectorized<float>::loadu(&in_ptr0[(16*i0) + (32*i1)]);
                    for (long i0_inner = 0; i0_inner < 16; i0_inner++) {
                        auto tmp1 = in_ptr1[i1 + (16*i0_inner) + (256*i0)];
                        tmp2[i0_inner] = tmp1 /* slice store */;
                    }
                    auto tmp3 = at::vec::Vectorized<float>::loadu(&tmp2[0]) /* register load */;
                    auto tmp4 = tmp0 + tmp3;
                    tmp4.store(out_ptr0 + (16*i0) + (32*i1));
                }
            }
            #pragma omp for 
            for(long i0=32; i0<32; i0+=1)
            {
                #pragma GCC ivdep
                for(long i1=0; i1<16; i1+=1)
                {
                    auto tmp0 = in_ptr0[i0 + (32*i1)];
                    auto tmp1 = in_ptr1[i1 + (16*i0)];
                    auto tmp2 = tmp0 + tmp1;
                    out_ptr0[i0 + (32*i1)] = tmp2;
                }
            }
        }
    }
}
```
### Limitations and next steps
* Cleanup `CppVecKernel`, `CppTile2DKernel` and `CppTile2DTailKernel` (in the next PR).
* Open vectorization for more kernels even if some of the ops cannot be vectorized.
* Explicit vectorization path (1D vec, 2D transpose) only supports tile sizes with HW vector length (8 or 16), otherwise falls back to scalar path. Next step: Extend the ops handler to support tile sizes of multiple HW vector length (maybe with an unrolling wrapper).
* Limited tiling pattern support: same as current master:
-- 1D tiling on inner-most loop level (both parallel and reduction) for most vec ops
-- 2D tiling on inner-most loop level and one outer loop level for transposition
Next step: support more tiling patterns, e.g., vertical vectorized reduction, vector dot etc.
* Heuristics to decide the right tiling sizes and patterns for a kernel

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

File: torch/_inductor/codegen/cpp.py [hunk 1]
```diff
@@ -1948,10 +2794,14 @@ def run(kernel):
...
-                    codegen_kernel(CppTile2DKernel, tiling_factor, outer_tiling_idx)
+                    codegen_kernel(
+                        CppTileKernel,
+                        [tiling_factor, tiling_factor],
+                        [outer_tiling_idx, inner_most_idx],
+                    )
                 )
                 inner_tail_loop.set_kernel(
-                    codegen_kernel(CppTile2DTailKernel, tiling_factor, outer_tiling_idx)
+                    codegen_kernel(CppTileKernel, [tiling_factor], [outer_tiling_idx])
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:2804]: `CppTile2DKernel` and `CppTile2DTailKernel` are not used any longer. Will remove them in the next PR to make the change simpler in this PR.

**Comment by jansel** [comment]
Any performance numbers?

**Comment by jgong5** [comment]
> Any performance numbers?

I don't expect performance changes with this PR since it is primarily a code refactoring. But I will make sure there is no perf regression.

**Comment by jansel** [comment]
> I don't expect performance changes with this PR since it is primarily a code refactoring. But I will make sure there is no perf regression.

If this is refactoring shouldn't it remove the old code it replaces?

**Comment by jgong5** [comment]
> > I don't expect performance changes with this PR since it is primarily a code refactoring. But I will make sure there is no perf regression.
> 
> If this is refactoring shouldn't it remove the old code it replaces?

Yes, I plan to do that in the next PR as I noted in the description in order to make the code review easier (otherwise, there would be irrelevant paired changeset showing up on the review). I can do cleanup in this PR if you prefer. Feel free to let me know.

**Comment by jgong5** [comment]
Will refactor and add optimizations gradually from another PR stack https://github.com/pytorch/pytorch/pull/97626. Close this stack for now.

**Pull request closed without merging**